Refactor: update heat source handling, introduce PyPS-onic pre-heating for geothermal and PTES

amos-schledorn opened the pull request:
Supersedes #1765 .

Following a discussion with @fneum, this feature is being rehauled and its description updated. For reference, this is the "old" description of its previous architecture:

Supersedes #1765 .

This feature implements some refactoring of how **Pit Thermal Energy Storage (PTES)** and **heat sources** are handled.
PTES boosting is now possible without custom constraints. Instead, energy balancing is handled more PyPS-onicly (yes, that's a word now!) through a few extra busses and links.

Geothermal boosting is introduced similarly with a preheater to better approximate real-world behaviour.
This results in an overall cleaner/more generic integration of heat sources.

## 1. New `HeatSource` Enum Class (`scripts/definitions/heat_source.py`)

A new enumeration class has been added to categorize heat sources:

| Heat Source | Properties |
|-------------|------------|
| **AIR, GROUND, SEA_WATER** | Inexhaustible sources - no resource bus needed |
| **GEOTHERMAL, RIVER_WATER** | Limited sources requiring a generator and resource bus |
| **PTES** | Special case - storage discharge treated as heat source |

Key properties for each source:
- `requires_bus`: Whether a resource bus is needed (True for limited sources)
- `requires_generator`: Whether a generator component is needed (geothermal, river_water)
- `has_constant_temperature`: Whether temperature is time-invariant (geothermal)

The class also provides methods for:
- Getting capital costs and lifetimes
- Determining heat pump bus2 connections
- Calculating efficiency2 for heat pumps (accounts for heat drawn from source)
- Generating carrier and bus names

## 2. `build_ptes_operations` Rule

**New configuration parameters:**
```yaml
sector:
  district_heating:
    ptes:
      enable: true
      temperature_dependent_capacity: false  # e_max_pu varies with ΔT
      charge_boosting_required: false  # NOT IMPLEMENTED
      discharge_resistive_boosting: false  # use resistive heaters vs heat pumps
      top_temperature: 90  # or "forward" for dynamic
      bottom_temperature: 10  # or "return" for dynamic
      design_top_temperature: 90
      design_bottom_temperature: 35
```

**Outputs:**
1. `temp_ptes_top_profiles`: PTES top layer temperature profile
2. `ptes_e_max_pu_profiles`: Normalized storage capacity (accounts for temperature-dependent capacity)
3. `ptes_boost_per_discharge_profiles`: Ratio of boost energy to discharge energy when T_forward > T_top. Used iff `sector:district_heating:ptes:discharge_resistive_boosting: true`

## 3. `build_heat_source_utilisation_profiles` Rule

Renamed from `build_heat_source_direct_utilisation_profiles`. Calculates when heat sources can be used and in what mode:

**Outputs:**
1. `heat_source_direct_utilisation_profiles`: Binary (1.0 when T_source ≥ T_forward for direct supply) - this already exists
2. `heat_source_preheater_utilisation_profiles`: Efficiency when T_return < T_source < T_forward (preheating mode) - this is new

## 4. Changes in `prepare_sector_network:add_heat`: new bus structure

### a) For exhaustible heat sources (currently geothermal, river-water and PTES with HP boosting but more to come in #1944)

```
Generator (only geothermal / river_water)
    │
    ▼
┌──────────────────┐
│   Resource Bus   │
└────────┬─────────┘
         │
         │  Utilisation Link
         │
    ┌────┴────────────────────────────┐
    │                                 │
    │ eff = 1 if T_source ≥ T_forward │ eff2 = 1 - eff
    │        else 0                   │
    ▼                                 ▼
urban                          Preheater-Input Bus
central                              │
heat                                 │  Preheater Link
                                     │
                ┌────────────────────┴────────────────────────┐
                │                                             │
                │ eff = T_source - T_return                            │ eff2 = 1 - eff
                │       ─────────────────────────             │
                │       (T_source - T_return) + ΔT_cooling    │
                ▼                                             ▼
            urban                                      Heat-Pump-Input Bus
            central                                          │
            heat                                             │
                                                             │  Heat Pump Link
                                                             │
                                        ┌────────────────────┴───┐
                                        ▲                        │
                                        │ eff = 1/COP            │ eff2 = 1 - 1/COP
                                        │                        ▼
                                Electricity                  urban
                                   Bus                       central
                                                             heat
```

For PTES, the generator is replaced with the `urban central water pits discharger` link.

### b) For Inexhaustible Sources (air / ground / sea_water)


**No buses created** — these sources don't need resource tracking, thus
- `heat_source.requires_bus` returns `False`
- `heat_source.get_heat_pump_input_bus()` returns `""`, which is also `bus2` for heat pumps.

**Links created:**

1. **Heat Pump Link**: Electricity → urban central heat
   - bus0: urban central heat (inverted link configuration)
   - bus1: Electricity Bus (eff = 1/COP)
   - bus2: not used (empty string)
   - No resource bus needed since the source is inexhaustible

```
                          Heat Pump Link
                                │
               ┌────────────────┴───┐
               ▲                    │
               │ eff = 1/COP        │ (no eff2 / bus2)
               │                    ▼
       Electricity              urban
          Bus                   central
                                heat
```

### c) PTES with Resistive Boosting

When `ptes:discharge_resistive_boosting: true`, PTES uses resistive heaters instead of heat pumps for the temperature lift. The structure bypasses the cascading utilisation/preheater/heat-pump chain.

**Key calculation (in `build_ptes_operations`):**
- **α = boost_per_discharge = (T_forward - T_top) / (T_top - T_bottom)**
- This is the ratio of boost energy needed relative to PTES discharge energy
- The `water pits discharger` link still points to the PTES resource bus `urban central ptes heat`
- Preheater and direct utilisation busses and links are created but `heat_source.requires_heat_pump()` returns `False`.
- Since no heat pump is created, all PTES heat is either routed through direct utilisation or resistive boosting, not through preheating.


```
                  ┌──────────────────┐
                  │   PTES Store     │
                  │   (water pits)   │
                  └────────┬─────────┘
                           │
                           │  Discharger Link                            
                          ▼
                  ┌─────────-────────┐
                  │   Resource Bus   │----> Utilisation Link -----> Urban central heat
                  │   (ptes heat)    │
                  └────────┬─────────┘
                         │
                         │  Resistive Booster Link
                         │
                        ┌────────────┴────────────────────┐
                        ▲                                 │
                        │ eff = α / (α + 1)               │ eff2 = 1 / (α + 1)
                        │                                 ▼
                 Resistive Heat                       urban
                      Bus  ──────────────┐            central
                        │                │            heat
                        │                │              ▲
                        │                │              │
           Resistive Heater        Stand-alone Link     │
           eff = 1                 eff = 1              │
                ▲                        │              │
                │                        └──────────────┘
         Electricity
            Bus
```

With resistive boosting, the ratio α determines how much of the output comes from PTES discharge vs electric heating:
- **eff2 = 1/(α+1)**: use 1 unit of PTES heat to produce α+1 units of district heating
- **eff = α/(α+1) = 1 - eff2**: energy conservation

## 5. Further onfiguration Changes

- Renamed `heat_pump_sources` → `heat_sources`
- Renamed `tes` → `ttes` (tank thermal energy storage)
- Removed `limited_heat_sources`, `direct_utilisation_heat_sources`, `temperature_limited_stores`


## Open issues
- [x] integration into brownfield setups
- [x] decide upon default behaviour (disable PTES/enable PTES with/without boosting; temperatures, ...)
- [x] add bus structure sketches to docs
- [x] calculation of pre-heating ratios needs reviewing

## Checklist

- [x] I tested my contribution locally and it works as intended.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in `config/config.default.yaml`.
- [x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] Sources of newly added data are documented in `doc/data_sources.rst`.
- [x] A release note `doc/release_notes.rst` is added.


## Testing

### Config

```run:
  name: 
  - ptes_90-10_dynamic_capacity_hpBoosting
  - ptes_fwd-ret_dynamic_capacity_hpBoosting
  - ptes_90-35_static_capacity_rhBoosting
  - ptes_90-10_dynamic_capacity_rhBoosting_geothermal
  - ptes_200-10_dynamic_capacity_hpBoosting
  scenarios:
    enable: true
    file: dev/scenarios.yaml
    shared_resources:
      policy: true

foresight: overnight

scenario:
  clusters:
  - 8
  planning_horizons:
  - 2040

clustering:
  temporal:
    resolution_sector: 25h

countries: ['DE', 'DK']

plotting:
  energy_threshold: 0
  enable_heat_source_maps: true
  interactive_bus_balance: 
    bus_name_pattern: "*urban central*"

sector:
  district_heating:
    supply_temperature_approximation:
      max_forward_temperature_baseyear:
        DE: 110
        DK: 80
      min_forward_temperature_baseyear:
        DE: 75
        DK: 55
      return_temperature_baseyear:
        DE: 60
        DK: 50
      relative_annual_temperature_reduction: 0.0
  ttes: false

##### scenarios.yaml #####


ptes_90-10_dynamic_capacity_hpBoosting:
  sector:
    heat_sources:
      urban central:
        - air
        - ptes
    district_heating:
      ptes:
        top_temperature: 90
        bottom_temperature: 10
        temperature_dependent_capacity: true
        discharge_resistive_boosting: false

ptes_fwd-ret_dynamic_capacity_hpBoosting:
  sector:
    heat_sources:
      urban central:
        - air
        - ptes
    district_heating:
      ptes:
        top_temperature: forward
        bottom_temperature: return
        temperature_dependent_capacity: true
        discharge_resistive_boosting: false

ptes_90-35_static_capacity_rhBoosting:
  sector:
    heat_sources:
      urban central:
        - air
        - ptes
    district_heating:
      ptes:
        top_temperature: 90
        bottom_temperature: 35
        temperature_dependent_capacity: false
        discharge_resistive_boosting: true

ptes_90-10_dynamic_capacity_rhBoosting_geothermal:
  sector:
    heat_sources:
      urban central:
        - air
        - ptes
        - geothermal
    district_heating:
      ptes:
        top_temperature: 90
        bottom_temperature: 10
        temperature_dependent_capacity: true
        discharge_resistive_boosting: true


ptes_200-10_dynamic_capacity_hpBoosting:
  sector:
    heat_sources:
      urban central:
        - air
        - ptes
    district_heating:
      ptes:
        top_temperature: 200
        bottom_temperature: 10
        temperature_dependent_capacity: true
        discharge_resistive_boosting: false
```
### Results

#### Parameters
(look reasonable)
<img width="1212" height="1160" alt="image" src="https://github.com/user-attachments/assets/80950f37-caba-4d59-8a23-1041bb248085" />


#### Example energy balance (Jan 3rd, `ptes_90-10_dynamic_capacity_hpBoosting`

##### `DE0 1 urban central ptes heat` bus
PTES discharge: 369 MWh

<img width="1126" height="621" alt="image" src="https://github.com/user-attachments/assets/a7e9ce8c-a01c-46fa-bf71-a3cf2b39cfe2" />

#####  `DE0 1 urban central ptes medium-temperature heat` bus
Discharge is fully absorbed by pre-heater, since no direct utilisation allowed

<img width="1174" height="680" alt="image" src="https://github.com/user-attachments/assets/3121da9e-463e-4cc2-b78b-f475695fbae2" />

#####  `DE0 1 urban central ptes return-temperature heat` bus
231 MWh is used as heat pump input
 
<img width="1174" height="680" alt="image" src="https://github.com/user-attachments/assets/21e4dc4d-36de-4ffd-be5f-5e9b9444c0a4" />

#####  `DE0 1 urban central heat` bus
Remaining 138 MWh are used by preheater to heat return flow
Heat pump output is 282 MWh > 231 MWh due to  electricity input

<img width="2718" height="1566" alt="image" src="https://github.com/user-attachments/assets/e909ddf2-5a70-4b3c-8095-579b3b0598c8" />

#### Entire year, `ptes_90-10_dynamic_capacity_hpBoosting`, `DE0 1`
Discharging over the entire year follows parameters shown above, with direct utilisation mostly in the summer.
There are some periods where both direct utilisation and pre-heating are happening due to the `25h` resolution.

<img width="2718" height="1566" alt="image" src="https://github.com/user-attachments/assets/0386c523-81e8-4aaa-84e3-87518bb73077" />

#### Entire year, `ptes_90-10_dynamic_capacity_hpBoosting`, `DK1 0`
Direct utilisation is possible throughout the entire year due to lower forward temperatures.

<img width="2718" height="1566" alt="image" src="https://github.com/user-attachments/assets/a5fb4fc4-898f-4ac1-8b72-d3716f478d42" />


#### Entire year, `ptes_90-10_dynamic_capacity_rhBoosting_geothermal`, `DE0 0`
Here, the pre-heating ratio for geothermal seems a little high but matches the logic, as `(T_source - T_return) / (T_source - T_return + 6K) = 0.45`.

<img width="2718" height="1566" alt="image" src="https://github.com/user-attachments/assets/f83ae883-64a6-4eb2-a5fc-5cd0c24d6217" />


## EDIT
I've also updated `build_geothermal_heat_source_potentials` by introducing scaling of Manz et al.'s data.
Manz et al. assume a difference of 15K. This rule now scales the potentials based our actual temperature delta:

a) If source_temperature > forward_temperature (direct utilisation):
   scale_factor = (source_temperature - return_temperature) / 15 K

b) If forward_temperature >= source_temperature > return_temperature (preheating):
   scale_factor = (source_temperature - return_temperature + heat_source_cooling) / 15 K

c) If source_temperature <= return_temperature (heat pump only):
   scale_factor = heat_source_cooling / 15 K

## EDIT 2
The example results are using a buggy scaling of `p_max_pu` for geothermal heating. I've fixed that bug resulting in the model not building goethermal heating anymore. I've checked that it can (by lowering CAPEX) and pre-heating ratios are still valid. 
(just a note that when executing these configs, geothermal has to be forced in to get comparable results, so I'm leaving the "faulty" results here)


## EDIT 3
Following a suggestion of @fneum's, I've compared a scenario without PTES & geothermal heating, i.e. those technologies that should be affected by the feature, to the master:
```
only_unaffected_technologies:
  sector:
    district_heating:
      ptes:
        enable: false
    heat_sources:
      urban central:
        - air
        - river_water
        - sea_water
    ttes: true
  solving:
    options:
      noisy_costs: false
```
On the master, I have removed the section in `prepare_sector_network` that builds PTES. 
-> The objective value is constant across both feature and master to the 5th digit:
```feature=1.16952e+11 master=1.16952e+11  (Δ=57755.6, +0.000%)```

I've also compared runs with default settings in the feature and master (still DE/DK-25h) with similar results (different colours though).

District heating balance in master, default settings:
<img width="903" height="425" alt="image" src="https://github.com/user-attachments/assets/a196c7f8-25cb-44aa-a496-e61dac57870d" />

and feature, default settings:
<img width="910" height="437" alt="image" src="https://github.com/user-attachments/assets/2741527f-3734-4af5-80a0-c5dc4c555898" />

Energy balance in master:
<img width="897" height="766" alt="image" src="https://github.com/user-attachments/assets/79f430ff-7018-480d-9d16-a929bb47b656" />

and feature:
<img width="906" height="759" alt="image" src="https://github.com/user-attachments/assets/42e1578c-e235-4302-8f2d-a59918a9737f" />

amos-schledorn committed:
refactor:introduct heat-source enum and pre-heating

amos-schledorn committed:
refactor:update PTES boosting rule

amos-schledorn committed:
update workflow

amos-schledorn committed:
core:update workflow

amos-schledorn committed:
introduce ptes design temperatures

amos-schledorn committed:
fix boosting bus naming

amos-schledorn committed:
fix resistive booster bus

amos-schledorn committed:
nicer colours

amos-schledorn committed:
disallow boosting when boosting_profile=0

amos-schledorn committed:
disable PTES heat pumps when discharge_resistive_boosting is True

amos-schledorn committed:
disallow preheating when preheater_utilisation=0

amos-schledorn committed:
Revert "disable PTES heat pumps when discharge_resistive_boosting is True"

This reverts commit ba2658804900c5418a37699847bb27989eb10d8d.

amos-schledorn committed:
fix PTES heat pump condition

amos-schledorn committed:
hot-fix bus structure for boosting

amos-schledorn committed:
rename intermediate boosting buses

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
Move bus/carrier naming to heat_source class and update colours

amos-schledorn committed:
update docs

amos-schledorn committed:
move heat generator down to surpress bus-not-defined warning

amos-schledorn committed:
update docs

amos-schledorn committed:
have geothermal require preheating

amos-schledorn committed:
set PTES heat-source cooling to T_top - T_bottom

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
Correct preheater profiles

amos-schledorn committed:
update test config

amos-schledorn committed:
update perfect/brownfield runs

amos-schledorn committed:
update geothermal to reflect dT assumed in data source

amos-schledorn committed:
update docs of build_geothermal_heat_potential

amos-schledorn committed:
update geothermal preheater colour

amos-schledorn committed:
fix myopic runs

amos-schledorn committed:
Merge branch 'master' into refactor-ptes-boosting

amos-schledorn committed:
fix bugs resulting from merge

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
update docs and release notes

amos-schledorn committed:
update default config

amos-schledorn committed:
fix: update heat source condition to check for geothermal type

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
fix heat source power input function

amos-schledorn committed:
add dummy colours for river_water util/preheating

amos-schledorn committed:
don't build resistive booster if ptes:enable:false

amos-schledorn committed:
handle case of empty heat sources

amos-schledorn committed:
enable ptes and geothermal as heat sources by default

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
fix: clarify preheating logic in get_sink_inlet_temperature function

amos-schledorn committed:
Merge branch 'master' into refactor-ptes-boosting

lindnemi commented on scripts/add_existing_baseyear.py:
```diff
@@ -580,40 +581,55 @@ def add_heating_capacities_installed_before_baseyear(
+                        ]
+                        * ratio,
+                        p_max_pu=0,
+                        p_min_pu=-1 * efficiency / efficiency.clip(lower=0.001),
```
this line could be simplified to `[-1 if eff > 0 else 0 for eff in efficiency]`

cpschau reviewed (changes requested):
Great work @amos-schledorn! I think PR description and the code complemented each other very well, which made it almost a joy to review. There are only very few and minor things to change & discuss before this is ready to merge!

amos-schledorn committed:
fix: update p_max_pu and p_min_pu calculations

amos-schledorn committed:
fix: update PTES temperature parameters to allow for forward_temperature/return_temperature

amos-schledorn committed:
fix: update parameter name from 'ttes' to 'ptes' in dynamic PTES capacity check

amos-schledorn committed:
docs: add notes on ideal heat exchangers to get_source_inlet_temperature function

amos-schledorn committed:
fix: update requires_preheater docs

amos-schledorn committed:
fix: add PTES enable validation in heat source utilisation profiles

amos-schledorn committed:
style: improve docstrings, move requires_heat_pump property upward

amos-schledorn commented on scripts/definitions/heat_source.py:
```diff
@@ -0,0 +1,425 @@
+        if self in [HeatSource.PTES, HeatSource.GEOTHERMAL]:
+            return True
+        else:
+            return False
```
I considered that but couldn't come up with a good solution for time-varying temperatures (sometimes above, sometimes below return temperature) and for users experimenting with different temperatures (e.g. excess temperatures around the return temperature). The structure should, however, handle unnecessary pre-heaters by setting efficiencies their to zero. 
(In fact, pre-heaters are always build for all heat sources but might not be used - I'm assuming solvers will handle that in pre-solve).
Happy to hear suggestions!

amos-schledorn committed:
fix: update parameter name from 'tes' to 'ttes' in docs

cpschau commented on doc/configtables/sector.csv:
```diff
@@ -20,12 +20,14 @@ district_heating,--,,
--- -- -- booster_heat_pump: true,--,"{true, false}",Add option to enable a booster heat pump for supplemental heating of pit storage in district heating
--- -- max_top_temperature,C,float,The maximum top temperature of the pit storage according to DEA technology catalogue (2018)
--- -- min_bottom_temperature,C,float,The minimum bottom temperature of the pit storage according to DEA technology catalogue (2018)
+-- -- enable,--,"{true, false}",Enable investments in pit thermal energy storage (PTES) in district heating
```
Agree about tackling reordering in a separate PR, but lets change the config table entry from `tes`to `ttes`, to align with the renaming of the config entry.

cpschau reviewed (changes requested):
Minor things remaining:

- config table key should mirror config keys
- dynamic COP source/sink inlet adjustments for boostable sources

amos-schledorn committed:
refactor: make preheating dynamic depending on temperature and rename medium/return temperature busses

amos-schledorn committed:
fix: disable discharge resistive boosting in default config

amos-schledorn committed:
fix: handle inexhaustible heat sources in get_heat_pump_efficiency2 method

amos-schledorn committed:
fix: transpose heat data in add_heat function for correct indexing

amos-schledorn committed:
fix incorrect preheater power limits and clean up obsolete method calls

amos-schledorn committed:
fix: update heat pump input bus method calls for consistency

amos-schledorn commented:
Thanks for the review, @cpschau. I have
- renamed `return_temperature_bus` and `medium_temperature_bus` to `heat_pump_input_bus` and `preheater_input_bus`
- removed the manual `requires_preheater` attribute: the code now handles this dynamically by routing all exhaustible heat sources 
- tested the updated workflow (not documenting results again) by comparing model dispatch to manually computed boosting and preheating requirements

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into merge-master-into-boosting

amos-schledorn committed:
update config and validation

amos-schledorn committed:
refactor: simplify heat sources validation in SectorConfig

amos-schledorn committed:
fix: remove unnecessary transpose in heat source data reindexing

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
fix: add ptes configuration for district heating in add_brownfield rule

amos-schledorn committed:
Merge branch 'master' into refactor-ptes-boosting

cpschau commented on scripts/prepare_sector_network.py:
```diff
@@ -3115,98 +3129,91 @@ def add_heat(
+                / 100,  # convert %/hour into unit/hour
+                capital_cost=costs.at["central water pit storage", "capital_cost"],
+                lifetime=costs.at["central water pit storage", "lifetime"],
+                e_nom_min=100000,
```
PTES fandom must not be lived by manipulating the upstream model parameters 😆

cpschau reviewed (changes requested):
Logic for sink inlet T must be changed, and e_nom_min for PTES must be removed.

amos-schledorn committed:
refactor: remove debugging e_nom_min for PTES

amos-schledorn committed:
fix incrrect sink-inlet temperature for COP approximation

amos-schledorn committed:
Merge branch 'merge-master-into-boosting' into refactor-ptes-boosting

amos-schledorn commented on scripts/build_cop_profiles/run.py:
```diff
@@ -106,6 +135,135 @@ def get_cop(
+    temperature. When preheating is not used (source <= return), the heat pump
+    receives water at return temperature and heats it to forward temperature.
+
+    When source temperature > return temperature, preheater is used: preheater raises return flow, heat pump inlet is at forward temp.
```
Good catch! I think it's actually supposed to be
```
return xr.where(
        source_temperature > central_heating_return_temperature,
        central_heating_return_temperature,
        source_temperature,
)
```
(which is what I implemented)

amos-schledorn committed:
refactor: discard error handling in get_source_temperature function

amos-schledorn committed:
docs: clarify scaling conditions for geothermal heat potentials

amos-schledorn committed:
docs: enhance docstring for resource_bus method in HeatSource class

amos-schledorn committed:
refactor: update PTES resource bus creation in add_heat function

amos-schledorn committed:
refactor: correct logic in get_sink_inlet_temperature function documentation and implementation

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
fix: handle dataarray conversion of heating technologies more robustly

amos-schledorn committed:
fix: improve error handling in logarithmic_mean method for temperature calculations

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

amos-schledorn committed:
Merge tag 'v2026.02.0' into refactor-ptes-boosting

v2026.02.0

amos-schledorn committed:
first draft: simplify boosting

amos-schledorn committed:
update boosting logic

amos-schledorn committed:
fix boosting logic:
1 utilisation link distributes to UCH and HP input

amos-schledorn committed:
fix faulty HP input for inexhaustible heat sources

amos-schledorn committed:
clean up docstrings

amos-schledorn committed:
fix util link ratios

amos-schledorn committed:
move util efficiencies to heat_source and fix util efficiency2

amos-schledorn committed:
fix resistive boosting, update schema and docstrings

amos-schledorn committed:
cleanup stale code, update docstrings

amos-schledorn committed:
Merge remote-tracking branch 'origin/master' into refactor-ptes-boosting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

amos-schledorn committed:
revert to source-side modelling

amos-schledorn committed:
Merge remote-tracking branch 'origin/simplify-boosting' into refactor-ptes-boosting

amos-schledorn committed:
fix merge artifacts

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
